File: src/main/java/xyz/phanta/cbtweaker/singleblock/SingleBlockMachineTileEntity.java
package xyz.phanta.cbtweaker.singleblock;

import io.github.phantamanta44.libnine.tile.L9TileEntityTicking;
import io.github.phantamanta44.libnine.tile.RegisterTile;
import io.github.phantamanta44.libnine.util.data.serialization.AutoSerialize;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import org.apache.commons.lang3.tuple.Pair;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.common.ActivatableMachine;
import xyz.phanta.cbtweaker.common.MachineDataHolder;
import xyz.phanta.cbtweaker.gui.inventory.UiElement;
import xyz.phanta.cbtweaker.util.block.MachineSideHandler;
import xyz.phanta.cbtweaker.util.block.RedstoneControlHandler;

import java.util.List;

@RegisterTile(CbtMod.MOD_ID)
public class SingleBlockMachineTileEntity extends L9TileEntityTicking implements ActivatableMachine {

    @AutoSerialize
    private final MachineDataHolder<SingleBlockData<?, ?, ?, ?, ?>> dataHolder = new MachineDataHolder<>(
            () -> SingleBlockData.construct(this,
                    ((SingleBlockMachineBlock)getWorld().getBlockState(getPos()).getBlock()).getSingleBlockType()));

    public SingleBlockMachineTileEntity() {
        markRequiresSync();
        setInitialized();
    }

    @Override
    protected ICapabilityProvider initCapabilities() {
        return getSideHandler();
    }

    public SingleBlockType<?, ?, ?> getSingleBlockType() {
        return dataHolder.getData().getSingleBlockType();
    }

    @Override
    public boolean isActive() {
        return dataHolder.getData().isActive();
    }

    public MachineSideHandler getSideHandler() {
        return dataHolder.getData().getSideHandler();
    }

    public RedstoneControlHandler getRedstoneHandler() {
        return dataHolder.getData().getRedstoneHandler();
    }

    public List<Pair<Object, UiElement>> createBufferUiElements() {
        return dataHolder.getData().createBufferUiElements();
    }

    public UiElement createRecipeLogicUiElement() {
        return dataHolder.getData().createRecipeLogicUiElement();
    }

    @Override
    public void setDirty() {
        super.setDirty();
    }

    public boolean handleInteraction(EntityPlayer player, EnumHand hand, EnumFacing face) {
        World world = getWorld();
        BlockPos pos = getPos();
        return dataHolder.getData().handleInteraction(world, pos, world.getBlockState(pos), player, hand, face);
    }

    public void dropContents() {
        dataHolder.getData().dropContents(getWorld(), getPos());
    }

    @Override
    protected void tick() {
        dataHolder.getData().tick();
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/singleblock/SingleBlockMachineTileEntity.kt
package st.evening.mc.cbtweaker.singleblock

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.ITickable
import net.minecraft.util.math.BlockPos
import net.minecraftforge.common.capabilities.Capability
import st.evening.mc.cbtweaker.common.LazyTileEntity
import st.evening.mc.cbtweaker.common.MachineTileEntity
import st.evening.mc.cbtweaker.gui.inventory.UiElement
import st.evening.mc.cbtweaker.util.TileEntitySyncProxy
import st.evening.mc.cbtweaker.util.component.RedstoneControlHandler
import st.evening.mc.cbtweaker.util.component.SidedBufferHandler
import st.evening.mc.cbtweaker.util.component.UiElementTable
import st.evening.mc.prelude.api.util.game.ServerSide

class SingleBlockMachineTileEntity : LazyTileEntity<SingleBlockData<*>>(), MachineTileEntity, ITickable {
    override fun initData(): SingleBlockData<*> {
        val data = SingleBlockData(this, (world.getBlockState(pos).block as SingleBlockMachineBlock).sbType)
        initSync(TileEntitySyncProxy(this, data.syncState))
        return data
    }

    val sbType: SingleBlockType<*>
        get() = data.sbType

    val bufHandler: SidedBufferHandler
        get() = data.bufHandler

    override val isActive: Boolean
        get() = data.isActive

    override val rsHandler: RedstoneControlHandler?
        get() = data.rsHandler

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean =
        bufHandler.hasCapability(capability, facing)

    override fun <T : Any> getCapability(capability: Capability<T>, facing: EnumFacing?): T? =
        bufHandler.getCapability(capability, facing)

    override fun update() {
        data.tick()
    }

    fun handleInteraction(
        player: EntityPlayer, hand: EnumHand, face: EnumFacing, hitX: Float, hitY: Float, hitZ: Float
    ): Boolean = data.handleInteraction(world.getBlockState(pos), player, hand, face, hitX, hitY, hitZ)

    @ServerSide
    fun handleBlockUpdate(blockState: IBlockState, fromBlock: Block, fromPos: BlockPos) {
        data.handleBlockUpdate(blockState, fromBlock, fromPos)
    }

    @ServerSide
    fun handleDestruction(blockState: IBlockState) {
        data.handleDestruction(blockState)
    }

    fun createBufferUiElements(): UiElementTable = data.createBufferUiElements()

    fun createMachineUiElement(): UiElement? = data.createMachineUiElement()
}
